Repeater Control

lukecarbis opened the pull request:
Closes #36.
Closes #391.

lukecarbis committed:
Add onboarding class which creates some dummy data

lukecarbis committed:
Remove max-height restriction on fieldRows

lukecarbis committed:
Add repeater control and area for sub fields to be populated

lukecarbis commented:
This is still _VERY_ early – it only includes 1 thing: An admin UI for adding / managing Repeater fields.

The "Add Sub-Field" button doesn't even work yet.

But I wanted to get an early review before heading too far down this path to make sure we're happy with the UX. What are your thoughts?

![first-pass-at-repeater](https://user-images.githubusercontent.com/1097667/56861091-1f961b80-69e1-11e9-8a92-bfdcf3142004.gif)

RobStino commented:
Hey @lukecarbis,
So excited to see this branch kicked off!!!!!
Here are the designs I mocked up for how I think this could look.
I'm not 100% committed to particular design bits and pieces, but I think it could be important to have the parent field persistently open and "wrapping" the sub fields. I recall we talked about borrowing from the (non-Customizer) menu builder for submenu items, but the above makes the sub-fields feel a bit disconnected from the parent field.
What do you think?
![Block_lab_Repeater_Field](https://user-images.githubusercontent.com/1578148/56863443-ab1ba680-69f9-11e9-8eb1-378205a22629.png)

RobStino commented:
Here is a mock of what the frontend could look like as well: https://user-images.githubusercontent.com/1578148/56863483-05b50280-69fa-11e9-9e28-cad86f719af2.png

lukecarbis commented:
I actually disagree. I think the backend gets far too crowded like that. Since the headers are all the same, isn't it better to keep all the fields visible without clicking into the repeater?

lukecarbis committed:
Add repeater control class

lukecarbis committed:
Add child rows when the Add Sub Field button is pressed

lukecarbis commented:
Added a commit which enables new Sub Fields to be added. There are still a few known issues (e.g. background colours when dragging, dragging sub-rows doesn't work, the auto-scroll doesn't go to the right place), but I just wanted to give you a little more of an idea on how this could work.

kienstra commented:
**Looks Good**

Hi @lukecarbis,
Nice, this looks good:

<img width="1002" alt="adding-sub-field" src="https://user-images.githubusercontent.com/4063887/56872348-b5d03e80-69ed-11e9-9fc9-42bf9be47895.png">

lukecarbis committed:
Use proper event delegation for the add child field button

RobStino commented:
Short screencap of initial thoughts on the WIP.
As discussed in today's team call.
https://d.pr/v/8N9Nrx

lukecarbis committed:
Prevent sub rows from collapsing when editing parent row in the block editor

lukecarbis committed:
Updating styling of repeater sub field rows

lukecarbis committed:
Add instructions for adding sub-fields

lukecarbis committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Updating styling of repeater sub field rows

lukecarbis committed:
Remove Add Sub Field helper text

lukecarbis committed:
Update styling of field rows

lukecarbis committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Merge branch 'feature/rich-text' into feature/repeater

lukecarbis committed:
Remove default value from the repeater

lukecarbis committed:
Fix bug which caused the child rows actions section to remain after changing control types

lukecarbis committed:
Rename child rows to sub rows

lukecarbis committed:
Prevent the possibility of adding a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, in a repeater, etc.

lukecarbis commented:
@kienstra I merged #221 into this PR, since I know that it's got some changes that affect the repeater field.

Progress so far: The UI for adding a repeater field _should_ now be done. I still haven't gotten around to saving it, and I need to do a general tidy up of the CSS. But those should be the last things to do for the block editor side of things (still lots to do in the editor editor).

@RobStino – could you please review the design / UX so far?

![2019-05-29 23-59-45 2019-05-30 00_20_40](https://user-images.githubusercontent.com/1097667/58564711-c8e66080-8270-11e9-9295-9358805baff3.gif)

lukecarbis committed:
Use settings config for the repeater help setting

pcfreak30 commented:
I think UI wise you may want to consider the possibility of a lot of repeater nesting. So I think if you add something, it may need to be a modal? Trying to add in 3-4 levels of repeaters would make the UI very cramped.

@lukecarbis

pcfreak30 commented:
I think commit 5775c6590c78b2319ed4522841ab8646d928be2c should actually be removed TBH. That flexibility should be allowed I think.

lukecarbis commented:
Thanks for your thoughts @pcfreak30. Although I agree that having multiple depths of nesting should be the goal – I think that for this release we should stick to just one. Otherwise we might never ship this massive feature.

lukecarbis committed:
Allow sub-fields to be saved as JSON

RobStino commented:
Hi @lukecarbis 
Great work on this so far.
I recall discussing moving the `location` field/column from the sub-fields table as it's not required.
Was this still the plan?

lukecarbis committed:
Slight change to phpunit install script

lukecarbis committed:
Add unit tests for the block, field, and repeater classes

lukecarbis committed:
Remove location from field settings

lukecarbis committed:
Fix subrow sorting for saved blocks

lukecarbis committed:
Modify unit test to be PHP 5.4 compatible

lukecarbis commented:
Nice pickup @Rob. Fixed in https://github.com/getblocklab/block-lab/pull/298/commits/b95782cb8fe7aaa69ab022e87d3930e943d3d0e3.

lukecarbis committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Whitespace fix

kienstra committed:
Merge in develop, resolve conflicts

There were conflicts in:
css/admin.block-post.css
php/blocks/class-field.php
php/post-types/class-block-post.php

Mainly resolve these conflicts in favor of this branch,
feature/repeater

RobStino commented:
@lukecarbis I was just looking at this and noticed that the "Location" column has been removed from the parent table all together. It's still available as a field option, but isn't visible in the table view. Was this intentional?

lukecarbis commented:
Yes, this is intentional. It's due to the fact that some fields don't have a location setting (e.g. Rich Text, Repeater, likely more advanced fields in the future).

I think we need a UI that addresses this in a different way.

kienstra committed:
Merge branch 'develop' into feature/repeater

kienstra commented:
**Question About Template Function**

Hi @lukecarbis and @RobStino,
Hope you had a great weekend.

How about a template function `block_sub_field( $parent, $child, $echo )` to output sub-field values in the template? This would essentially be a wrapper for [block_field()](https://github.com/getblocklab/block-lab/blob/a48846506c74342454150c0bb7c04d5f700f4f2d/php/helpers.php#L20), but for sub-fields. 

For example, in a block template like `blocks/block-home-carousel.php`:

```php
<img src="<?php block_sub_field( 'home-carousel', 'image-1' ); ?>">
```

ACF [has iteration functions](https://www.advancedcustomfields.com/resources/get_sub_field/) like `have_rows()` and `the_row()`. 

We won't have multiple nested repeater fields, just a parent repeater and sub-fields. So I don't see much need for those, but I'm open to ideas.

Thanks!

kienstra commented:
Ah, maybe multiple nesting is planned, per https://github.com/getblocklab/block-lab/pull/298#issuecomment-497277092

>Although I agree that having multiple depths of nesting should be the goal – I think that for this release we should stick to just one. Otherwise we might never ship this massive feature.

kienstra commented:
If there will eventually be multiple nesting, maybe the template function could accept an `array()` of the nested repeaters:

```php
get_sub_field( array( 'home-carousel', 'image-1', 'alt-1' ) )
```

That's getting complex, though. 

I'd like to avoid [iterator functions](https://www.advancedcustomfields.com/resources/get_sub_field/) like `the_row()` and `have_rows()` that use and set the global state.

lukecarbis commented:
@kienstra I'm actually all for keeping things as simple as possible. I have trouble imagining use cases that rely on complicated systems with multiple nestings of repeaters. I say we focus on doing a single level of repeater nesting, and doing it as simply as possible.

kienstra commented:
Thanks, @lukecarbis! That sounds great to me.

kienstra committed:
Resolve merge conflict from before that I probably committed

This was in the style rule:
#block_fields .block-fields-rows .block-fields-row-columns
Use this branch (feature-repeater), instead of the develop branch.
That rule is for calculated values in grid-template-columns.

RobStino commented:
Simple sounds good to me.
My only thoughts re this is that whether a decision to keep it simple locks us away from being able to expand it later on, if we chose to.
We may never do that, but keep it simple for now is a different decision to keeping it simple for ever. As this is an untried feature for our users, I don't know if we have the insights to make a forever decision.
If we can do simple now and keep it open to expanding in the future, that's ideal I think.

kienstra committed:
Refactor functions to enable rendering repeater fields

This is the beginning of the work to render
the repeater fields in Gutenberg.
It still needs a way to save these fields,
as they don't persist on clicking 'Update'
and reloading the page.

Also, the actual display of the repeater in Gutenberg
is still WIP.

WebGardenGeek commented:
Just read through this thread and I really like what you guys came up with. It's great that the repeater is essentially a container for whatever subfields you want to add.

I have used a plugin called WCK (WordPress Creation Kit Pro) for custom post type and options page development and they do something very similar for metabox fields, which when it comes right down to it are fairly similar to blocks, except they're not drag and drop and they are assigned to a post type instead of just being a universal block that one can add to a post, page, or admin options page etc.

You simply indicate whether you want a repeater or not and then you add variables to the repeater. Then when you go to actually edit that metabox on the actual post/page it allows you to fill out the variables and add them to the array and you can do that as many times as you want (the repeater part).

Seems like things are coming along - let me know if you want a beta tester ;-)

kienstra committed:
Add a JS repeater component, first working example of repeater

There is more work needed,
including more work on the UI.
And there isn't a separate function for the
repeater.

kienstra commented:
**Repeater Field In Editor**

This will probably need better styling and more thought about the layout. But here's the working Repeater field in the editor.

If you had created a block with a Repeater before, you'll probably have to go back to `/wp-admin` > Block Lab > Edit Block and click 'Update.'

<img width="933" alt="repeater-field" src="https://user-images.githubusercontent.com/4063887/61166910-aac08100-a4fb-11e9-8c08-43b96f276957.png">

Still, this doesn't yet have template functions to output the values on the front-end, like `block_field()`.

kienstra commented on js/blocks/controls/repeater.js:
```diff
@@ -0,0 +1,25 @@
+import { getRenderedFields } from '../loader/edit';
+
+const BlockLabRepeaterControl = ( props, field, block ) => {
+	const subFields = field.sub_fields ? getRenderedFields( field.sub_fields, props, block ) : null;
```
If it's alright, I changed `sub-fields` to `sub_fields`, to enable accessing it like `field.sub_fields`.

kienstra committed:
Give block_field() a way to output sub-fields

Allow passing an array() with the parent and sub-field.
For example, for a sub-field 'image',
block_field( 'home-hero', 'image' );

kienstra commented:
**Template function now works**

This is definitely open for discussion.

But https://github.com/getblocklab/block-lab/pull/298/commits/3c6b624117c1f7b0f30000e1b86b621022839d8d gives a way for `block_field()` to output repeater fields.

For example:
```php
block_field( 'home-carousel', 'image-1' )
```

The function still works like before.

Update: I'll probably have to reconsider this.

kienstra committed:
Remove extra empty line

In response to a failed PHP Travis build:
Functions must not contain multiple empty lines in a row;

kienstra commented:
I still have quite a bit of work in this PR, but Repeater fields are now working.

kienstra commented:
**Should the Repeater be available in the Inspector?**

Hi @lukecarbis,
Do you think the Repeater should be available in the Inspector? It might be hard to have a 'nested' effect, [like here](https://github.com/getblocklab/block-lab/pull/298#issuecomment-511086383). 

There isn't much room in the Inspector.

Thanks, Luke! Hope you're doing great.

kienstra committed:
Abstract logic into set_attributes_from_field()

Now, this can apply to both fields
and sub-fields.

kienstra committed:
Add another assertion to maybe_get_sub_field_name()

Test for there being a third, unnecessary field.
This had the correct first 2 fields.

kienstra commented on php/blocks/class-loader.php:
```diff
@@ -168,23 +168,43 @@ public function get_block_attributes( $block ) {
+	 * @param Field  $field      The Field to set the attributes from.
+	 * @return array $attributes The attributes, with the new field value set.
+	 */
+	public function set_attributes_from_field( $attributes, $field_name, $field ) {
```
This is only abstracted from `get_block_attributes()`, and not changed much.

kienstra commented:
**Request For Help With Block Editor Display**

Hi @lukecarbis and @RobStino,
Update: This probably doesn't need feedback until I implement the repeater rows in the block editor.

Could you please help with the layout and design of the Repeater field in the block editor?

Here's the current display:
<img width="1050" alt="current-repeater" src="https://user-images.githubusercontent.com/4063887/61189554-b2e00400-a654-11e9-8753-3ef0daba227e.png">

Feel free to commit to this PR, like to [repeater.js](https://github.com/getblocklab/block-lab/pull/298/files#diff-4c5cb88a2e8c8ccdc511d166dd025c8eR15). Or to share thoughts or screenshots.

Thanks!

kienstra committed:
Change comments and DocBlocks

Remove @todo that called for a recursion limit.
We still might have to revisit that,
but I'm not sure the function itself
needs to have a recursion limit.

kienstra committed:
Prevent nesting repeaters inside repeaters

This might eventually allow that,
but for now, prevent selecting
a repeater field inside a repeater.

lukecarbis committed:
Merge branch 'develop' into feature/repeater

kienstra committed:
Merge in develop, resolve conflict

The conflict was in class-block-post.php

Resolve in favor of the develop branch.
Also, fix the failing unit test
test_from_json().

kienstra committed:
Remove duplicate notice for Pro license

Somehow, it looks like I duplicated
this on resolving the merge conflict.

kienstra committed:
Enable adding and removing rows

There is still much remaining,
including a better uuid() function.
This assigns each row a unique number.
Then, it stores the fields using that number.
For example, for the field 'baz-text',
it might be a24df26e[baz-text].

kienstra commented:
**Current Display**

<img width="991" alt="repeater-here" src="https://user-images.githubusercontent.com/4063887/61605062-7d658880-ac09-11e9-9831-80559ae2560d.png">

* The repeater field now allows adding and removing rows.
* But it doesn't save its values in the block editor. I need to work on that.
* It probably needs styling and layout work.

kienstra commented:
**Question About Maximum Rows**

What do you think about having a field for 'Maximum Rows':

<img width="957" alt="maximum-rows" src="https://user-images.githubusercontent.com/4063887/61605755-958ad700-ac0c-11e9-851c-8290ced45023.png">

This would be the maximum amount of rows that a user could create in the Repeater field.

Ideally, this wouldn't be needed. But it looks like we'll have to declare the repeater row fields as block attributes, even before they're added.

The Repeater control allows dynamically adding rows ([shown above](https://github.com/getblocklab/block-lab/pull/298#issuecomment-513629775)). Each of them has fields, which become attributes of the block.

But it looks like blocks don't allow dynamically adding attributes. The attributes must be declared once in [registerBlockType](https://github.com/getblocklab/block-lab/blob/a3ee65d05e46f80552bd49198743a299dee624ef/js/blocks/loader/index.js#L37).

So if there were 2 sub-fields in the repeater:

```js
example-image
example-text
```

...there would need to be attributes declared in [registerBlockType](https://github.com/getblocklab/block-lab/blob/a3ee65d05e46f80552bd49198743a299dee624ef/js/blocks/loader/index.js#L37), even if the user hasn't added any row yet:

```js
{
        example-image[0]: { type: 'string' },  // For the first row.
        example-text[0]: { type: 'string' },
        example-image[1]: { type: 'string' }, // For the second row.
        example-text[1]: { type: 'string' },
        example-image[2]: { type: 'string' }, // For the third row.
        example-text[2]: { type: 'string' },
        // etc...
}
```

There probably needs to be some limit to this, it probably shouldn't declare hundreds of rows.

**That's why maybe there should be a 'Maximum Rows' field.**

I'm looking at other workarounds for this.

RobStino commented:
Layout idea as covered in today's product call.
Concept for moving away from stacking rows.
![image](https://user-images.githubusercontent.com/1578148/61675710-cc600a80-ad3c-11e9-9f8e-91064867b690.png)

lukecarbis committed:
Merge branch 'develop' into feature/onboarding

lukecarbis committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Merge branch 'feature/repeater' into feature/onboarding

lukecarbis committed:
Add onboarding styles

lukecarbis committed:
Update default category to array style

lukecarbis committed:
Add welcome message

lukecarbis committed:
Add a post-script anti-nag message to the onboarding intro

lukecarbis committed:
Use the terminology notice instead of message

lukecarbis committed:
Add CTA to edit the dummy post

lukecarbis committed:
Make strings translatable

lukecarbis committed:
Add onboarding for adding a field and adding the block to a post

lukecarbis committed:
Messaging updates

lukecarbis committed:
Delete onboarding option once onboarding is complete

lukecarbis committed:
Minor style fix for the template location in the list table

lukecarbis committed:
Move example post id check back to admin notices method

lukecarbis committed:
Restore get_post call in add fields notice

lukecarbis committed:
Only delete the example post id option once the post is published

lukecarbis committed:
Add message prompting the user to publish the block

lukecarbis committed:
Style the repeater rows a little

lukecarbis committed:
Update the inline commenting asterix style

lukecarbis committed:
Only check for 1 existing block post, instead of -1

lukecarbis committed:
Change string quotes style for the template locate in onboarding

kienstra committed:
Update the active fieldset, using state

The product of working together with Luke
on this.
There is still quite a bit of work needed here,
including moving this Component into components/.

kienstra committed:
Move the RepeaterFields component into components/

There aren't many changes,
though the activeFieldset state
is changed to activeRows

kienstra committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Move delete option to after the notice is shown, so that its not deleted during save

lukecarbis commented:
Note: I found a bug where including a space in the sub-field name causes the repeater to break. e.g., try creating a repeater with the sub-field "Button Text". You can even just set the slug to `button`.

Edit: No, I think the bug is just that saving a block without editing the repeater causes it to break.

lukecarbis committed:
Remove pointers and update the publish onboarding notice

lukecarbis committed:
Fix remove rows and pagination

lukecarbis committed:
Change row actions IconButtons or Buttons

lukecarbis committed:
Add a single row by default

lukecarbis committed:
Fix broken repeater when saving a block

lukecarbis commented:
@kienstra Back over to you mate. In terms of next steps, could we move the "Next" and "Previous" buttons outside of the individual rows, and alongside the "Add new" button?

![Screen Shot 2019-08-12 at 12 18 18 pm](https://user-images.githubusercontent.com/1097667/62843205-75c26c80-bcfb-11e9-850d-7db1804cd2df.png)

lukecarbis committed:
Merge branch 'develop' into feature/onboarding

kienstra committed:
Copy the set_util() method into init(), add helper method for post type

To follow the pattern of the init() method,
this moves the contents of set_util() into
that method, and deletes the method.
Also, it adds a helper method
get_post_type_slug() to Plugin.
This might remove the need to add $block_post
as a property of Plugin.

kienstra committed:
Replace another reference to ->block_post->slug with helper method

I missed this in the previous commit,
so replace it with get_post_type_slug().

kienstra committed:
Move get_post_type_slug() to from Plugin_Abstract Util

As Luke mentioned, Plugin_Abstract isn't the
best place for this,
as it should apply to any plugin.
So move it to Utils.

kienstra committed:
Fix failed unit tests by making a property public and changing setUp()

The setUp method should actually just
call ->set_plugin(),
instead of register_component().

kienstra committed:
Merge pull request #384 from getblocklab/add/init-update-feature-onboarding

Copy the set_util() method into init(), add helper method for post type

kienstra committed:
Begin moving components into components/ but no edits to actual components,

The repeater control needs the Fields component,
which the Edit component also needs.
Maybe there's a better way to organize this,
as in Gutenberg-type projects,
there are usually several components/ directories.
But there are shared components/ here,
so maybe starting with a single components/ is good.

kienstra committed:
Remove named default exports

Instead of naming the default exports,
simply export them.

kienstra committed:
More refactoring into components/, still WIP

Move loader/inspector to:
components/block-lab-inspector.js,
as it's not a component.
In general, anything returning JSX is going
in components/,
other than the actual controls.

kienstra committed:
Assign the components to a const before exporting

This should make it easier to debug,
as before, the console didn't reference
the name of the component.

kienstra committed:
Minor whitespace edits, including space inside ${ }

This should make it a little more
similar to the eslint coding style
that Gutenberg and related plugins use.

kienstra committed:
Merge branch 'develop' into feature/repeater

kienstra committed:
Move the previous and next buttons out of each row

Before, I had put a button for this in every row.
So move them outside of the rows.

lukecarbis committed:
Restore tests to exclude rules in phpcs (for now)

lukecarbis committed:
Merge branch 'develop' into feature/repeater

lukecarbis committed:
Merge branch 'feature/repeater' into feature/onboarding

lukecarbis committed:
Store default value data for repeater sub fields

lukecarbis committed:
Use new Util class for locate_template

lukecarbis committed:
Merge pull request #269 from getblocklab/feature/onboarding

Add onboarding class which creates some dummy data

lukecarbis committed:
Remove fixed height with scroll for blocks in the editor

lukecarbis committed:
Remove maybe_get_sub_field_name helper function

lukecarbis committed:
Stub out helper functions for sub fields

kienstra committed:
Add DocBlocks for dependencies, like Internal dependencies

Following the patter in Gutenberg,
use DocBlocks for internal and WordPress dependencies.

lukecarbis committed:
Add Loop class and loop helper functions

lukecarbis committed:
Move typecasting from block_field to the Field class

lukecarbis committed:
Refactor block_field to resolve undefined index bugs

lukecarbis committed:
Add active property to loop class

lukecarbis committed:
First pass at block_sub_field

lukecarbis committed:
Minor docblock ammendment

kienstra committed:
Allow the repeater to save the row fields

Add an onChange handler to the <Fields> component.
This will handle changing repeater rows,
or normal fields.
There's still more work, including
possibly refactoring the rest of the controls.

kienstra committed:
Make remaining Controls a little more like React components

Make them accept only props as an argument,
and get the properties from destructuring.
This allows passing more to it,
it already had 3 arguments as a traditional function.

kienstra committed:
Fix the Post control call to the ContentControl

The props caused an error,
so move them to a contentProps constant.

lukecarbis committed:
Merge branch 'feature/repeater' into feature/repeater-helpers

kienstra committed:
Fix an issue with the onChange const not being present

As Luke found,
this was not present in the Textarea Control.

lukecarbis committed:
Merge branch 'feature/repeater' into feature/repeater-helpers

lukecarbis committed:
Update sub field helper functions to match repeater data storage

kienstra committed:
Fix the getValue() function, remove a few extra constants

The getValue() constant wasn't working.
Also, remove some constants that aren't needed
with earlier refactoring.

lukecarbis committed:
Update repeater styles, and move editor.scss to css/src

kienstra committed:
Minor updates to the Loop class

Also, add a cast helper method.
Simply the previous logic,
extracted to a helper method.

kienstra committed:
Update the DocBlock summary of cast_value_to_string()

I should have changed this in
the previous commit, where the function
name changed.

kienstra committed:
Merge pull request #387 from getblocklab/feature/repeater-helpers

Helper functions for the repeater fields

lukecarbis committed:
Check if field is set before calling a Field method

kienstra committed:
Move ContentControl to components/

This isn't a compenent that's registered on its own,
it's simply used by other components.

lukecarbis committed:
Add styles for repeater rows

lukecarbis committed:
Fix broken removeRow method

lukecarbis committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater

kienstra committed:
Move content.js to content-control.js to avoid errror

I should have moved this in the previous commit,
so move it here to avoid a build error.

lukecarbis committed:
Conditionally show move up and down arrows in repeater

lukecarbis committed:
Style the new row button

lukecarbis committed:
Move rows up and down

kienstra committed:
Stop using the parentBlockProps directly in most controls

This gave the access to values from many components
higher in the tree.
Instead, mainly rely on getValue() and
onChange().
The Repeater control will still use the parentBlockProps, though.

kienstra committed:
Remove the onChange prop from Fields

This wasn't used, though it was in place
to possibly handle the repeater.

kienstra committed:
Simplify the content controls

One of them, the Post control,
had a needless function: getTitleFromAPI().

lukecarbis committed:
Don't show Location setting for sub-fields

lukecarbis committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater

kienstra committed:
Fix an issue with the Inspector control

Before, it wasn't saving properly,
as it was calling the controls with multiple arugments,
not with props.

kienstra committed:
Fix an issue with adding new rows

Before, I had this using the onChange() handler.
But instead, use the parentBlockProps.

kienstra commented on js/blocks/controls/post.js:
```diff
@@ -1,15 +1,7 @@
-	const getTitleFromAPI = apiResponse => ( apiResponse && apiResponse.title && apiResponse.title.rendered ) ? apiResponse.title.rendered : '';
-
-	return ContentControl( props, field, getTitleFromAPI );
+const BlockLabPostControl = ( props ) => {
```
It looks like `getTitleFromAPI()` wasn't actually used in the `ContentControl`, or anywhere else. I'm not sure why I added it 😄

kienstra commented on js/blocks/loader/edit.js:
```diff
@@ -1,67 +0,0 @@
-	)
-}
-
-const editComponent = ( props, block ) => {
```
This file is mainly refactored. For example, `editComponent` is moved to `components/edit.js`

kienstra committed:
Improve the naming of a function I had named

Change set_attributes_from_field()
to get_attributes_from_field,
as it gets a value.

kienstra committed:
Fix an issue with the Taxonomy control

Before, it didn't display the selected taxonomy.
Now, it seems to work.

kienstra committed:
Allow the Range control to get the value

Before, getValue wasn't available.
So simply add it to the destructuring line.

kienstra committed:
Minor DocBlock changes, and adding ;

Change some of the @param tags,
like from {Number} to {number},
to match the convention used in Gutenberg and AMP.

kienstra committed:
Minor edits, like removing parameter from getParent()

It simply uses a property of the component,
so get it from the property instead.

kienstra committed:
More minor edits, like enclosing a parameter in ( )

To follow the convention in Gutenberg and AMP.
Also, other minor edits like adding { } for an if block.

kienstra committed:
Merge in develop, resolve conflict.

There was a conflict in admin.block-post.js
I retained both edits,
as they seemed to just be 2 functions
added in the same place.

kienstra committed:
Remove a ; that caused a failed build

I added this in,
but it looks like it causes a failed npm run dev

lukecarbis committed:
Use post type slug util method instead of this->slug

lukecarbis committed:
Store the post type slug in a parameter

kienstra committed:
Allow saving non-repeater fields

Before, in /wp-admin > Block Lab > Edit Block,
it wasn't possible to save non-repeater fields.
On clicking 'Update', any non-repeater fields
didn't save.
It looks like this is because $field->settings['parent']
was set, but empty.
So instead, check for it being empty.

kienstra committed:
Fix an issue in displaying the post title

It looks like I had broken this,
so it actually displayed the post ID,
instead of the title.
So allow it to show the title.

lukecarbis committed:
Add transition animation to moving repeater rows

lukecarbis committed:
Add get url from path util function

kienstra commented on js/blocks/components/advanced-controls.js:
```diff
@@ -15,15 +24,15 @@ const inspectorControls = ( props, block ) => {
+				<p className="bl-dot-tip" dangerouslySetInnerHTML={ { __html: tip } }></p>
 				<p className="bl-dot-tip read-more">
-					<a href="https://github.com/getblocklab/block-lab/wiki/7.-FAQ" target="_blank">{__( 'Read more', 'block-lab' )}</a>
+					<a href="https://github.com/getblocklab/block-lab/wiki/7.-FAQ" target="_blank">{ __( 'Read more', 'block-lab' ) }</a>
```
@todo: update this link

kienstra committed:
Remove unused code, where it was looking for a nonexistent controlFunction property

Now, it will simply look at the controls from controls/index.js
This is the previous behavior.

kienstra commented on js/blocks/components/fields.js:
```diff
@@ -0,0 +1,82 @@
+			const attr = { ...parentBlockProps.attributes };
+			const { setAttributes } = parentBlockProps;
+
+			if ( undefined !== rowIndex ) { // If this is in a repeater row.
```
@todo: change order of if/else

kienstra replied:
This is applied in https://github.com/getblocklab/block-lab/pull/298/commits/5abc5644bf97c1123e5864b30a7b6993ca332902

kienstra commented on js/blocks/components/fields.js:
```diff
@@ -0,0 +1,82 @@
+			const attr = { ...parentBlockProps.attributes };
+
+			// If the field has a parent, it's probably in a repeater row.
+			return ( field.parent && attr[ field.parent ][ rowIndex ] ) ? attr[ field.parent ][ rowIndex ][ field.name ] : attr[ field.name ];
```
@todo: Make this a proper if statement.

kienstra replied:
This is applied in https://github.com/getblocklab/block-lab/pull/298/commits/a0a0ac2f98801cbfb8be4d1ee80bd2270c8858be.

kienstra committed:
Remove the <div> wrapper of <Fields> in <FormControls>

It looks like there's no need for that,
as it's a simple <div>.

lukecarbis committed:
Move block post slug definition to block post class

kienstra committed:
Remove the wrapper around Control

This doesn't look to be needed.
Also, make some more minor edits.

kienstra committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater

kienstra committed:
Revert "Move block post slug definition to block post class"

This reverts commit 6ddab72ff58a76d0b17a223c246eb02b9b5f0781.

kienstra committed:
Break a ternary expression into an if/else block

As Luke mentioned,
this is easier to understand.

kienstra committed:
Make an if/else block easier to read

As Luke mentioned,
it'll be easier to have undefined === rowIndex
than undefined !== rowIndex.

kienstra committed:
Remove ControlContainer, simply place in Fields component

This was only used in one place,
so it's not very necessary to have a separate file.

lukecarbis committed:
Update Upgrade screen to remove references to Early Access

lukecarbis committed:
Update documentation URLs

lukecarbis committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater

kienstra committed:
If the parent doesn't exist in the block attributes, exit

In block_sub_field(),
there was an issue where calling a non-existent

field name produced an error.

lukecarbis committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater

kienstra committed:
Fix an issue where the Repeater's default values didn't display

There was a workaround that added default values,
but it didn't apply to sub-fields.

Props Luke Carbis.

lukecarbis committed:
Update to the upgrade page

kienstra committed:
Store the rows in a rows attribute of an object

Before, the repeater attribute was an array.
This will allow more flexibility,
as the repeater can save more values in the future.

lukecarbis committed:
Merge branch 'feature/repeater' of https://github.com/getblocklab/block-lab into feature/repeater
